From 5bdffe3481d2e47773bb35824a2a682573369676 Mon Sep 17 00:00:00 2001
From: Fei Jiang <fei.jiang@intel.com>
Date: Fri, 29 Dec 2017 18:48:22 +0800
Subject: [PATCH 0485/1240] drm/i915/gvt: add param disable_gvt_fw_loading to
 disable gvt fw loading

when gvt fw doesn't exist, try to load fw from initrd will cost about 1 minute,
add disable_gvt_fw_loading to speed up sos boot up time.

Change-Id: I07aca795057fec09ff9ee729a65a5f96006c5ef8
Signed-off-by: Fei Jiang <fei.jiang@intel.com>
Reviewed-by: Dong, Eddie <eddie.dong@intel.com>
Tested-by: Dong, Eddie <eddie.dong@intel.com>
---
 drivers/gpu/drm/i915/gvt/firmware.c | 12 ++++++++----
 drivers/gpu/drm/i915/i915_params.c  |  4 ++++
 drivers/gpu/drm/i915/i915_params.h  |  3 ++-
 3 files changed, 14 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/gvt/firmware.c b/drivers/gpu/drm/i915/gvt/firmware.c
index a26c1705..9f8a0b6 100644
--- a/drivers/gpu/drm/i915/gvt/firmware.c
+++ b/drivers/gpu/drm/i915/gvt/firmware.c
@@ -220,10 +220,6 @@ int intel_gvt_load_firmware(struct intel_gvt *gvt)
 	void *mem;
 	int ret;
 
-	path = kmalloc(PATH_MAX, GFP_KERNEL);
-	if (!path)
-		return -ENOMEM;
-
 	mem = kmalloc(info->cfg_space_size, GFP_KERNEL);
 	if (!mem) {
 		kfree(path);
@@ -241,6 +237,13 @@ int intel_gvt_load_firmware(struct intel_gvt *gvt)
 
 	firmware->mmio = mem;
 
+	if (i915.disable_gvt_fw_loading)
+		goto expose_firmware;
+
+	path = kmalloc(PATH_MAX, GFP_KERNEL);
+	if (!path)
+		return -ENOMEM;
+
 	sprintf(path, "%s/vid_0x%04x_did_0x%04x_rid_0x%02x.golden_hw_state",
 		 GVT_FIRMWARE_PATH, pdev->vendor, pdev->device,
 		 pdev->revision);
@@ -248,6 +251,7 @@ int intel_gvt_load_firmware(struct intel_gvt *gvt)
 	gvt_dbg_core("request hw state firmware %s...\n", path);
 
 	ret = request_firmware(&fw, path, &dev_priv->drm.pdev->dev);
+
 	kfree(path);
 
 	if (ret)
diff --git a/drivers/gpu/drm/i915/i915_params.c b/drivers/gpu/drm/i915/i915_params.c
index 1632f34..c945a65 100644
--- a/drivers/gpu/drm/i915/i915_params.c
+++ b/drivers/gpu/drm/i915/i915_params.c
@@ -68,6 +68,7 @@ struct i915_params i915 __read_mostly = {
 	.enable_pvmmio = 1,
 	.enable_gvt_oos = 1,
 	.enable_conformance_check = true,
+	.disable_gvt_fw_loading = true,
 };
 
 module_param_named(modeset, i915.modeset, int, 0400);
@@ -271,3 +272,6 @@ MODULE_PARM_DESC(enable_gvt_oos, "To toggle the gvt ppgtt page table OOS (Out of
 
 module_param_named(enable_conformance_check, i915.enable_conformance_check, bool, 0400);
 MODULE_PARM_DESC(enable_conformance_check, "To toggle the GVT guest conformance feature.");
+
+module_param_named(disable_gvt_fw_loading, i915.disable_gvt_fw_loading, bool, 0400);
+MODULE_PARM_DESC(disable_gvt_fw_loading, "Disable GVT-g fw loading.");
diff --git a/drivers/gpu/drm/i915/i915_params.h b/drivers/gpu/drm/i915/i915_params.h
index 33d2fe3..a81515f 100644
--- a/drivers/gpu/drm/i915/i915_params.h
+++ b/drivers/gpu/drm/i915/i915_params.h
@@ -71,7 +71,8 @@
 	func(bool, enable_gvt); \
 	func(bool, enable_pvmmio); \
 	func(bool, enable_gvt_oos); \
-	func(bool, enable_conformance_check)
+	func(bool, enable_conformance_check); \
+	func(bool, disable_gvt_fw_loading);
 
 #define MEMBER(T, member) T member
 struct i915_params {
-- 
2.7.4

